Dev

linkevinlin1 opened the pull request:
First working versions of batch, stream, and producers

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+import json
+from datetime import datetime, timedelta
+
+kafka_servers = 'b-1.kafka.0ofh61.c4.kafka.us-east-1.amazonaws.com:9092,b-2.kafka.0ofh61.c4.kafka.us-east-1.amazonaws.com:9092'
```
Don't hardcode addresses!  (everywhere else as well)

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def kafka_init(servers,topic):
+    '''
```
Irrelevent, but I prefer PEP style triple quotes """ instead of '''
Also, you can use type hints (https://www.python.org/dev/peps/pep-0484/) instead of having a separate @type line

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+playback = 24 # n times
+window = 20  # in minutes
+convert_win = int(window * 60 / 24)
+lookbackdays = 5
```
parameters like these might be more appropriate to load in from a config file (.txt, JSON, whatever)

also they should have more detailed comments describing usage

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+# output timestamp is at the middle of the window for next day
+
+
+def kafka_init(servers,topic):
```
You reuse kafka_init and acked in different scripts, which means they should be loaded in from a single separate file

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+    Used for kafka message acknowledgement
+    '''
+    if err is not None:
+        print("Failed to deliver message: %s: %s" % (str(msg), str(err)))
```
Replace ALL print statements with logging (https://docs.python.org/3/library/logging.html) (good in general but extra important for streaming projects)
One advantage is that logging has different "levels" so you can put error messages like this at a higher level than routine steps, or redirect low level logs to a file but print out important errors on screen.

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+producer = kafka_init(kafka_servers, topic)
+
+starttime=time.time()
+while True:
```
The while True: steps here and elsewhere should be replaced with `if __name__ == "main"` executing a function

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+while True:
+
+    then = datetime.now()
+    utc_timestamp = int(time.mktime(then.timetuple())*1e3 + then.microsecond/1e3 - convert_win * 1000 / 2 + 86400 * 1000 / playback)
```
this should maybe be a function.

I'm wondering if you should have a separate file(s) for functions relating to time manipulations

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+
+    for i in range(lookbackdays):
+        now = now - timedelta(seconds = 86400 / playback)
+        interval_list[i] = now.replace(microsecond=0).isoformat() + '/pt' + str(convert_win*2) + 's'
```
this line should maybe be a function

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+        post_aggregations={'avg_power': (Field('sum_power_w') / Field('count_w'))}
+    )
+
+    df = query.export_pandas()
```
it seems like there should be some way to wrap up the column deletions into the query instead of exporting then removing.

colindiemer commented on batch/druid_batch.py:
```diff
@@ -0,0 +1,101 @@
+
+    for i, row in df.iterrows():
+        value_out = row.to_json()[:-1]+','+json.dumps(time_json)[1:]
+        producer.produce(topic, key='key', value = value_out, callback=acked)
```
Instead of iterows can you use more manipulations on the dataframe first and use something like .to_json?

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+    'building6': ['plasma TV', 'lamp', 'toaster', 'hob', 'iron', 'computer w/ scanner and printer', 'LCD TV', 'washing machine', 'fridge w/ freezer'],
+    'building7': ['hair dryer', 'washing machine', 'videogame console and radio', 'dryer', 'TV w/ decoder and computer in living room', 'kitchen TV', 'dishwasher', 'total outlets', 'total lights'],
+    'building8': ['kitchen TV', 'dishwasher', 'living room TV', 'desktop computer w/ screen', 'washing machine', 'bedroom TV', 'total outlets', 'total lights'],
+    }
```
good for config file

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+
+def house_reader_GREEND(datapath):
+    '''
+    Load the GREEND files into iterators.
```
it would be good if somewhere in the code there is documentation as to the schema of these files

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+    @rtype filenames: list[iterator]
+    @return: file name iterator for each house
+    '''
+    path = datapath+'/GREEND'
```
Personal preference, but I don't like hard coded path manipulations (what if I want use `/greend` instead?). e.g. you can just use datapath and a config file.

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+    '''
+    path = datapath+'/GREEND'
+    house_list = sorted(os.listdir(path))
+    house_list = [house for house in house_list if "building" in house]
```
Good usage of list comprehensions throughout :)

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+    filenames = [[] for _ in range(len(house_list))]
+    for hid, house in enumerate(house_list):
+        hpath = path + '/' + house
+        filenames[hid]=iter(sorted([os.path.basename(x) for x in glob.glob(hpath+"/dataset_201*.csv")]))
```
If you're using a lot of filename manipulations it might make your life easier to work with something like pathlib (https://docs.python.org/3/library/pathlib.html) instead of string manipulations

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+prevalues = [None]*len(house_list)
+# csv iterator for row reading
+csviters = [None]*len(house_list)
+
```
it might be cleaner to do these with multi-assignment, e.g. like `(x,y,z) = ["whatever"]*3`

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+                prevalues[house_id] = [float(x) if is_number(x) else 0 for x in entries[1:]]
+            else:
+                pretimes[house_id] = prevalues[house_id] = None
+
```
at least some of these steps need to be wrapped in functions for clarity.

One general principle is if there are multiple lines following a single `if` statement, you might consider creating a function. It's sometime much easier to read  `for i in thing: descriptive_function_name(i)` then to have the actual code block for the function there

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+    #print(filename)
+    csviter = csv.reader(fd, delimiter=',')
+    next(csviter, None)
+    return fd, csviter
```
I like the concept, and I'm big on functions in general, but this seems unnecessary, and just doing `file_descriptor = open()` and `csv_iter = csv.reader()` seems fine.

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+        if not pretimes[house_id]:
+            # close opened file
+            if fileds[house_id]:
+                fileds[house_id].close()
```
when working with `open()` in python it is better to use it in a `with open() as file` syntax, see 7.2 of https://docs.python.org/3/tutorial/inputoutput.html. This way you don't have to worry about closing. 

(Although you may have some more subtle reason for doing it your way I missed)

colindiemer commented on ingestion/producer_GREEND_batch.py:
```diff
@@ -0,0 +1,169 @@
+                pretimes[house_id] = float(entries[0])*1000
+                prevalues[house_id] = [float(x) if is_number(x) else 0 for x in entries[1:]]
+            else:
+                pretimes[house_id] = prevalues[house_id] = None
```
again, try to think of how to wrap some of these steps in functions for clarity

linkevinlin1 commented:
How do I add more changes to this pull request?
I forgot to add Spark Structured Streaming script and I fixed some of the problems in the producers, added Airflow to control the batch processing.

Edit: They seemed to be reflected here after I pushed them. Colin, if you got some time, please take a quick look on the streaming script. Thanks.

colindiemer commented:
> How do I add more changes to this pull request?
> I forgot to add Spark Structured Streaming script and I fixed some of the problems in the producers, added Airflow to control the batch processing.
> 
> Edit: They seemed to be reflected here after I pushed them. Colin, if you got some time, please take a quick look on the streaming script. Thanks.

Yeah, PRs automatically get updated with new commits into the same branch.

I'll try to take a quick look later today or tomorrow. 😄

linkevinlin1 commented:
Thanks a lot! Working on scaling things up right now. I'll get to other issues you mentioned later.
